### Importing Libraries

In [1]:
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns
import numpy as np

In [2]:
# Prevents rows from being truncated 
pd.options.display.max_columns = None
pd.options.display.max_rows = None


### Import data 

In [3]:
df = pd.read_csv('car_data_cleaned.csv')

In [4]:
df.head(20)

Year                  Model   Mileage                Location    Price  \
0   2011      335is Convertible   62,515             Somerset, NJ  $25,999   
1   2011  335i xDrive Coupe AWD   37,739               Easton, PA  $25,998   
2   2011             335i Sedan   67,840          Stroudsburg, PA  $18,990   
3   2010  335i xDrive Sedan AWD  128,507    Hasbrouck Heights, NJ   $9,995   
4   2011       335i Convertible   55,578               Easton, PA  $23,998   
5   2011       335i Convertible   71,447          Stroudsburg, PA  $21,990   
6   2010       335i Convertible   74,799          Stroudsburg, PA  $18,990   
7   2011      335is Convertible   40,262          Stroudsburg, PA  $28,990   
8   2011  335i xDrive Coupe AWD   34,908          Jersey City, NJ  $19,998   
9   2008       335i Convertible  137,855    Huntingdon Valley, PA  $11,295   
10  2009       335i Convertible   60,798          Stroudsburg, PA  $20,990   
11  2011       335i Convertible   89,479            Elizabeth, NJ  $17,985   
12  2011             335i Sedan  106,062             Garfield, NJ  $13,499   
13  2011      335is Convertible   62,515             Somerset, NJ  $25,999   
14  2011  335i xDrive Coupe AWD   37,739               Easton, PA  $25,998   
15  2011             335i Sedan   67,840          Stroudsburg, PA  $18,990   
16  2010  335i xDrive Sedan AWD  128,507    Hasbrouck Heights, NJ   $9,995   
17  2011       335i Convertible   55,578               Easton, PA  $23,998   
18  2011       335i Convertible   71,447          Stroudsburg, PA  $21,990   
19  2010       335i Convertible   74,799          Stroudsburg, PA  $18,990   

    Site Price Type                        Color Scheme  \
0        High Price     Silver exterior, Black interior   
1   Excellent Price      White exterior, Brown interior   
2       Great Price       Blue exterior, Brown interior   
3   Excellent Price      Black exterior, Beige interior   
4        High Price    Black exterior, Unknown interior   
5        High Price      Black exterior, Brown interior   
6        High Price      Black exterior, Black interior   
7        High Price       Blue exterior, Black interior   
8   Excellent Price     Blue exterior, Unknown interior   
9        High Price      Black exterior, Beige interior   
10       High Price      Black exterior, Brown interior   
11       High Price       Blue exterior, Beige interior   
12       High Price  Unknown exterior, Unknown interior   
13       High Price     Silver exterior, Black interior   
14  Excellent Price      White exterior, Brown interior   
15      Great Price       Blue exterior, Brown interior   
16  Excellent Price      Black exterior, Beige interior   
17       High Price    Black exterior, Unknown interior   
18       High Price      Black exterior, Brown interior   
19       High Price      Black exterior, Black interior   

                                 History 335i_yn xdrive_yn is_model_yn  \
0   No accidents, 3 Owners, Personal use     yes        no         yes   
1   No accidents, 2 Owners, Personal use     yes       yes          no   
2      No accidents, 5 Owners, Fleet use     yes        no          no   
3   No accidents, 3 Owners, Personal use     yes       yes          no   
4   No accidents, 2 Owners, Personal use     yes        no          no   
5   No accidents, 6 Owners, Personal use     yes        no          no   
6   No accidents, 3 Owners, Personal use     yes        no          no   
7   No accidents, 3 Owners, Personal use     yes        no         yes   
8   No accidents, 2 Owners, Personal use     yes       yes          no   
9   No accidents, 5 Owners, Personal use     yes        no          no   
10  No accidents, 3 Owners, Personal use     yes        no          no   
11    1 accident, 2 Owners, Personal use     yes        no          no   
12  No accidents, 5 Owners, Personal use     yes        no          no   
13  No accidents, 3 Owners, Personal use     yes        no         yes   
14  No accident

In [5]:
# Questions I want to answer:

# How much do 335is cost on average?
# Price compare between E90 & E92
# Do interior and exterior colors influence price

### Removing '$' and ',' from Price amd ',' from Mileage for EDA

In [6]:
df['Price'] = df['Price'].apply(lambda x: x.replace(',', '').replace('$',''))

In [7]:
df['Mileage'] = df['Mileage'].apply(lambda x: x.replace(',', ''))

### Convert 'Mileage' and 'Price' from object to int

In [8]:
df['Mileage'] = df['Mileage'].astype(int)

In [9]:
df['Price'] = df['Price'].astype(int)

### Preping variables for EDA

In [10]:
df['high_mileage'] = df['Mileage'].apply(lambda x: 'yes' if x > 120000 else 'no')

In [11]:
df['low_mileage'] = df['Mileage'].apply(lambda x: 'yes' if x < 90000 else 'no')

In [12]:
df['highmiles_int'] = df['high_mileage'].map(dict(yes=1, no=0))

In [13]:
df['lowmiles_int'] = df['low_mileage'].map(dict(yes=1, no=0))

In [14]:
df['xdrive'] = df['xdrive_yn'].map(dict(yes=1, no=0))

In [15]:
df['is_model'] = df['is_model_yn'].map(dict(yes=1, no=0))

In [16]:
df['accident_int'] = df['accident'].map(dict(Yes=1, No=0))

In [17]:
df.interior_color.value_counts()

 Brown       120
 Black        90
 Beige        90
 Unknown      90
Name: interior_color, dtype: int64

In [18]:
df = df.drop(['335i_yn','History', 'Color Scheme'], axis=1)

In [19]:
df.head(50)

Year                  Model  Mileage                Location  Price  \
0   2011      335is Convertible    62515            Somerset, NJ  25999   
1   2011  335i xDrive Coupe AWD    37739              Easton, PA  25998   
2   2011             335i Sedan    67840         Stroudsburg, PA  18990   
3   2010  335i xDrive Sedan AWD   128507   Hasbrouck Heights, NJ   9995   
4   2011       335i Convertible    55578              Easton, PA  23998   
5   2011       335i Convertible    71447         Stroudsburg, PA  21990   
6   2010       335i Convertible    74799         Stroudsburg, PA  18990   
7   2011      335is Convertible    40262         Stroudsburg, PA  28990   
8   2011  335i xDrive Coupe AWD    34908         Jersey City, NJ  19998   
9   2008       335i Convertible   137855   Huntingdon Valley, PA  11295   
10  2009       335i Convertible    60798         Stroudsburg, PA  20990   
11  2011       335i Convertible    89479           Elizabeth, NJ  17985   
12  2011             335i Sedan   106062            Garfield, NJ  13499   
13  2011      335is Convertible    62515            Somerset, NJ  25999   
14  2011  335i xDrive Coupe AWD    37739              Easton, PA  25998   
15  2011             335i Sedan    67840         Stroudsburg, PA  18990   
16  2010  335i xDrive Sedan AWD   128507   Hasbrouck Heights, NJ   9995   
17  2011       335i Convertible    55578              Easton, PA  23998   
18  2011       335i Convertible    71447         Stroudsburg, PA  21990   
19  2010       335i Convertible    74799         Stroudsburg, PA  18990   
20  2011      335is Convertible    40262         Stroudsburg, PA  28990   
21  2011  335i xDrive Coupe AWD    34908         Jersey City, NJ  19998   
22  2008       335i Convertible   137855   Huntingdon Valley, PA  11295   
23  2009       335i Convertible    60798         Stroudsburg, PA  20990   
24  2011       335i Convertible    89479           Elizabeth, NJ  17985   
25  2011             335i Sedan   106062            Garfield, NJ  13499   
26  2011      335is Convertible    62515            Somerset, NJ  25999   
27  2011  335i xDrive Coupe AWD    37739              Easton, PA  25998   
28  2011             335i Sedan    67840         Stroudsburg, PA  18990   
29  2010  335i xDrive Sedan AWD   128507   Hasbrouck Heights, NJ   9995   
30  2011       335i Convertible    55578              Easton, PA  23998   
31  2011       335i Convertible    71447         Stroudsburg, PA  21990   
32  2010       335i Convertible    74799         Stroudsburg, PA  18990   
33  2011      335is Convertible    40262         Stroudsburg, PA  28990   
34  2011  335i xDrive Coupe AWD    34908         Jersey City, NJ  19998   
35  2008       335i Convertible   137855   Huntingdon Valley, PA  11295   
36  2009       335i Convertible    60798         Stroudsburg, PA  20990   
37  2011       335i Convertible    89479           Elizabeth, NJ  17985   
38  2011             335i Sedan   106062            Garfield, NJ  13499   
39  2011      335is Convertible    62515            Somerset, NJ  25999   
40  2011  335i xDrive Coupe AWD    37739              Easton, PA  25998   
41  2011             335i Sedan    67840         Stroudsburg, PA  18990   
42  2010  335i xDrive Sedan AWD   128507   Hasbrouck Heights, NJ   9995   
43  2011       335i Convertible    55578              Easton, PA  23998   
44  2011       335i Convertible    71447         Stroudsburg, PA  21990   
45  2010       335i Convertible    74799         Stroudsburg, PA  18990   
46  2011      335is Convertible    40262         Stroudsburg, PA  28990   
47  2011  335i xDrive Coupe AWD    34908         Jersey City, NJ  19998   
48  2008       335i Convertible   137855   Huntingdon Valley, PA  11295   
49  2009       335i Convertible    60798         Stroudsburg, PA  20990   

    Site Price Type xdrive_yn is_model_yn body_style car_color interior_color  \
0        High Price        no         yes        E92    Silver         Black    
1   Excellent Pr

### Average cost of N54 335I's
#### E90 - \$14,000
#### E92 - \$21,000

In [24]:
pd.pivot_table(df,index=['body_style'], values=['Price'])

Price
body_style              
E90         14161.333333
E92         21623.300000

### Average cost of high mileage cars

#### High mileage E90's: \$9,995

#### High mileage E92's: \$11,295

In [27]:
pd.pivot_table(df,index=['body_style','highmiles_int'], values=['Price'])

Price
body_style highmiles_int              
E90        0              16244.500000
           1               9995.000000
E92        0              22770.888889
           1              11295.000000

### Average cost of low mileage cars

#### Low mileage E90's: \$18,990

#### Low mileage E92's: \$22,770

In [28]:
pd.pivot_table(df,index=['body_style','lowmiles_int'], values=['Price'])

Price
body_style lowmiles_int              
E90        0             11747.000000
           1             18990.000000
E92        0             11295.000000
           1             22770.888889

### All-Wheel drive cars vs non all wheel drive

In [29]:
pd.pivot_table(df,index=['body_style','xdrive'], values=['Price'])

Price
body_style xdrive           
E90        0       16244.500
           1        9995.000
E92        0       21279.625
           1       22998.000